File: src/big_config/step_fns.clj
(ns big-config.step-fns
  (:require [big-config.core :as core]
            [big-config.keys :as k]
            [big-config.utils :as u]
            [clojure.java.io :as io])
  (:import [java.nio.file Files]))

(defn exit-with-code [n]
  (System/exit (int n)))

(defn create-exit-step-fn [end]
  (core/create-step-fn
   {:after-f (fn [step opts]
               (when (and (= step end) (not (#{:repl "repl"} (get opts k/env))))
                 (exit-with-code (get opts k/exit 0))))}))

(def ->exit-step-fn create-exit-step-fn)

(defn create-print-error-step-fn [end]
  (core/create-step-fn
   {:before-f (fn [step opts]
                (let [err (get opts k/err)
                      exit (get opts k/exit)]
                  (when (and (= step end) (pos? (or exit 0)))
                    (binding [*out* *err*]
                      (when (and (string? err) (seq (.trim err)))
                        (println "✖" err))
                      (let [stack (get opts k/stack-trace)]
                        (when (and (string? stack) (seq (.trim stack)))
                          (let [dir (.toFile (Files/createTempDirectory "big-config-" (make-array java.nio.file.attribute.FileAttribute 0)))
                                file (io/file dir "stack-trace.txt")]
                            (spit file stack)
                            (println "\nThe stack-trace has been written to" (.getPath file)))))))))}))

(def ->print-error-step-fn create-print-error-step-fn)

(defonce taps (atom []))

(def tap-step-fn
  (core/create-step-fn
   {:before-f (fn [step opts] (swap! taps conj [step :before opts]))
    :after-f (fn [step opts] (swap! taps conj [step :after opts]))}))

(defn log-step-fn [f step opts]
  (f step (assoc opts k/steps-trace (conj (vec (get opts k/steps-trace [])) step))))

(def bling-step-fn
  (core/create-step-fn
   {:before-f (fn [step _opts]
                (binding [*out* *err*]
                  (println "➜" step)))
    :after-f (fn [step opts]
               (when (pos? (get opts k/exit 0))
                 (binding [*out* *err*]
                   (println "✖" (str step ":") (or (get opts k/err) "")))))}))

(u/register-function 'big-config.step-fns/tap-step-fn tap-step-fn)
(u/register-function 'big-config.step-fns/log-step-fn log-step-fn)
(u/register-function 'big-config.step-fns/bling-step-fn bling-step-fn)

File: src/big_config/step_fns.py
from __future__ import annotations

import sys
import tempfile
from typing import Any, Callable

from selmer import render as selmer_render

from . import ENV, ERR, EXIT, STACK_TRACE, STEPS
from .core import step_fn


def exit_with_code(n: int | None) -> None:
    sys.exit(0 if n is None else int(n))


def exit_step_fn(end: str):
    return step_fn(
        {
            "after_f": lambda step, opts: exit_with_code(opts.get(EXIT))
            if step == end and opts.get(ENV) != "repl"
            else None
        }
    )


def print_error_step_fn(end: str):
    def before(step: str, opts: dict[str, Any]) -> None:
        exit_code = opts.get(EXIT, 0)
        err = opts.get(ERR)
        stack_trace = opts.get(STACK_TRACE)
        if step == end and isinstance(exit_code, int) and exit_code > 0 and isinstance(err, str) and err.strip():
            print(f"✘ {err}", file=sys.stderr)
        if step == end and isinstance(exit_code, int) and exit_code > 0 and isinstance(stack_trace, str) and stack_trace.strip():
            with tempfile.NamedTemporaryFile("w", prefix="big-config-", suffix=".txt", dir="/tmp", delete=False) as handle:
                handle.write(stack_trace)
            print(f"\nThe stack-trace has been written to {handle.name}", file=sys.stderr)

    return step_fn({"before_f": before})


def _tap(label: str, step: str, opts: dict[str, Any]) -> None:
    # Python has no tap> equivalent here; keep a hook-friendly no-op.
    return None


tap_step_fn = step_fn({"before_f": lambda step, opts: _tap("before", step, opts), "after_f": lambda step, opts: _tap("after", step, opts)})


def log_step_fn(f: Callable[[str, dict[str, Any]], dict[str, Any]], step: str, opts: dict[str, Any]) -> dict[str, Any]:
    new_opts = dict(opts)
    new_opts[STEPS] = [*new_opts.get(STEPS, []), step]
    return f(step, new_opts)


def bling_step_fn(f: Callable[[str, dict[str, Any]], dict[str, Any]], step: str, opts: dict[str, Any]) -> dict[str, Any]:
    print(selmer_render("{{ prefix }} {{ msg }}", {"prefix": "→", "msg": str(step)}), file=sys.stderr)
    result = f(step, opts)
    if result.get(EXIT, 0) > 0:
        print(selmer_render("{{ prefix }} {{ msg }}: {{ err }}", {"prefix": "✘", "msg": str(step), "err": result.get(ERR)}), file=sys.stderr)
    return result


# Compatibility aliases.
globals()["->exit-step-fn"] = exit_step_fn
globals()["->print-error-step-fn"] = print_error_step_fn
